src/sage/modules/with_basis/representation.py

```diff
@@ -278,13 +350,127 @@ def representation_matrix(self, g, side=None, sparse=False):
                 temp = g * B[k]
             else:
                 temp = B[k] * g
-            for m, c in temp._monomial_coefficients.items():
+            for m, c in temp.monomial_coefficients(copy=False).items():
                 if not use_left:
                     ret[i, inv_order[m]] = c
                 else:
                     ret[inv_order[m], i] = c
         return ret
 
+    @cached_method
+    def character(self):
+        r"""
+        Return the character of ``self``.
+
+        EXAMPLES::
+
+            sage: SGA = SymmetricGroupAlgebra(QQ, 5)
+            sage: SM = SGA.specht_module([3,2])
+            sage: SM.character()
+            (5, 1, 1, -1, 1, -1, 0)
+            sage: matrix(SGA.specht_module(la).character() for la in Partitions(5))
+            [ 1  1  1  1  1  1  1]
+            [ 4  2  0  1 -1  0 -1]
+            [ 5  1  1 -1  1 -1  0]
+            [ 6  0 -2  0  0  0  1]
+            [ 5 -1  1 -1 -1  1  0]
+            [ 4 -2  0  1  1  0 -1]
+            [ 1 -1  1  1 -1 -1  1]
+
+            sage: SGA = SymmetricGroupAlgebra(QQ, SymmetricGroup(5))
+            sage: SM = SGA.specht_module([3,2])
+            sage: SM.character()
+            Character of Symmetric group of order 5! as a permutation group
+            sage: SM.character().values()
+            [5, 1, 1, -1, 1, -1, 0]
+            sage: matrix(SGA.specht_module(la).character().values() for la in reversed(Partitions(5)))
+            [ 1 -1  1  1 -1 -1  1]
+            [ 4 -2  0  1  1  0 -1]
+            [ 5 -1  1 -1 -1  1  0]
+            [ 6  0 -2  0  0  0  1]
+            [ 5  1  1 -1  1 -1  0]
+            [ 4  2  0  1 -1  0 -1]
+            [ 1  1  1  1  1  1  1]
+            sage: SGA.group().character_table()
+            [ 1 -1  1  1 -1 -1  1]
+            [ 4 -2  0  1  1  0 -1]
+            [ 5 -1  1 -1 -1  1  0]
+            [ 6  0 -2  0  0  0  1]
+            [ 5  1  1 -1  1 -1  0]
+            [ 4  2  0  1 -1  0 -1]
+            [ 1  1  1  1  1  1  1]
+        """
+        G = self._semigroup
+        B = self.basis()
+        chi = [sum((g * B[k])[k] for k in B.keys())
```

Comment:
perhaps
```suggestion
        chi = [sum((g * b)[k] for k, b in B.items())
```

Reply:
I prefer the explicitness here, and I would like to keep it the way it is.

Reply:
ok